gpiozero/mixins.py

```diff
@@ -577,9 +579,11 @@ def fill(self):
             while not self.stopping.wait(self.sample_wait):
                 value = self.parent._read()
                 if value not in self.ignore:
```

Comment:
Hmmm, if the `value` _is_ in `self.ignore`, then I guess `self.queue` won't get appended to, and therefore `len(self.queue)` won't change, and therefore there's no chance of `self.full` getting set either?
I.e. perhaps this function could change to:
```python
    def fill(self):
        try:
            while not self.stopping.wait(self.sample_wait):
                value = self.parent._read()
                if value not in self.ignore:
                    with self.lock:
                        self.queue.append(value)
                        if not self.full.is_set() and len(self.queue) >= self.queue.maxlen:
                            self.full.set()
                    if (self.partial or self.full.is_set()) and isinstance(self.parent, EventsMixin):
                        self.parent._fire_events(self.parent.pin_factory.ticks(), self.parent.is_active)
        except ReferenceError:
            # Parent is dead; time to die!
            pass
```
?

Reply:
Moving the middle if in the function inside the first lock (and eliminating the second `with self.lock:`) looks like a good change.

I'm not sure about that last if in the function being protected by `if value not in self.ignore:`

The existing code allows `self.parent._fire_events()` to be activated when `value` is in `self.ignore` and `self.partial` is set.  The suggested change won't allow that to happen.

Thoughts?